From 2c7722faab10d5d17b2aff690e5b91c80b6f540a Mon Sep 17 00:00:00 2001
Message-Id: <2c7722faab10d5d17b2aff690e5b91c80b6f540a.1779960237.git.keller@fs-net.de>
In-Reply-To: <bb92f3f011b7b125a61ac948b576f518cee03bb3.1779960237.git.keller@fs-net.de>
References: <bb92f3f011b7b125a61ac948b576f518cee03bb3.1779960237.git.keller@fs-net.de>
From: Hartmut Keller <keller@fs-net.de>
Date: Wed, 27 May 2026 18:07:14 +0200
Subject: [PATCH 19/20] Support newer kernel versions up to 6.6

Add support for newer Linux kernels up to at least 6.6.

- Add linux/ prefix in front of some included headers
- Add checks on LINUX_VERSION_CODE where newer kernels differ
---
 CORE/CLD_TXRX/TXRX/ol_txrx_internal.h       |   2 +-
 CORE/HDD/src/wlan_hdd_assoc.c               |   2 +-
 CORE/HDD/src/wlan_hdd_cfg80211.c            |  14 +++-
 CORE/MAC/src/include/parserApi.h            |   2 +-
 CORE/MAC/src/include/utilsApi.h             |   2 +-
 CORE/SME/src/csr/csrApiRoam.c               |  13 +++-
 CORE/SYS/legacy/src/system/inc/sysDebug.h   |   2 +-
 CORE/SYS/legacy/src/utils/inc/utilsParser.h |   2 +-
 CORE/SYS/legacy/src/utils/src/logApi.c      |   2 +-
 CORE/UTILS/PKTLOG/linux_ac.c                | 105 ++++++++++++++++++++++++++++
 CORE/VOSS/inc/vos_trace.h                   |   2 +-
 CORE/VOSS/src/vos_utils.c                   |   4 +-
 CORE/WDA/src/wlan_qct_wda_debug.h           |   2 +-
 13 files changed, 139 insertions(+), 15 deletions(-)

diff --git a/CORE/CLD_TXRX/TXRX/ol_txrx_internal.h b/CORE/CLD_TXRX/TXRX/ol_txrx_internal.h
index 73bb08e..ae8545b 100644
--- a/CORE/CLD_TXRX/TXRX/ol_txrx_internal.h
+++ b/CORE/CLD_TXRX/TXRX/ol_txrx_internal.h
@@ -106,7 +106,7 @@ extern unsigned g_txrx_print_level;
 
 #ifdef TXRX_PRINT_ENABLE
 
-#include <stdarg.h>       /* va_list */
+#include <linux/stdarg.h>       /* va_list */
 #include <adf_os_types.h> /* adf_os_vprint */
 
 /* Supress 4296 - expression is always true
diff --git a/CORE/HDD/src/wlan_hdd_assoc.c b/CORE/HDD/src/wlan_hdd_assoc.c
index 3142721..dc45c0e 100644
--- a/CORE/HDD/src/wlan_hdd_assoc.c
+++ b/CORE/HDD/src/wlan_hdd_assoc.c
@@ -6454,7 +6454,7 @@ static int __iw_get_ap_address(struct net_device *dev,
     }
     else
     {
-        memset(wrqu->ap_addr.sa_data,0,sizeof(wrqu->ap_addr.sa_data));
+        memset(wrqu->ap_addr.sa_data,0,ETH_ALEN);
     }
     EXIT();
     return 0;
diff --git a/CORE/HDD/src/wlan_hdd_cfg80211.c b/CORE/HDD/src/wlan_hdd_cfg80211.c
index dfd14fb..247d9f5 100644
--- a/CORE/HDD/src/wlan_hdd_cfg80211.c
+++ b/CORE/HDD/src/wlan_hdd_cfg80211.c
@@ -29935,7 +29935,14 @@ static int wlan_hdd_cfg80211_tdls_mgmt(struct wiphy *wiphy,
 					u16 status_code, u32 peer_capability,
 					const u8 *buf, size_t len)
 #else /* TDLS_MGMT_VERSION2 */
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 17, 0)) || defined(WITH_BACKPORTS)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 5, 0)) || defined(WITH_BACKPORTS)
+static int wlan_hdd_cfg80211_tdls_mgmt(struct wiphy *wiphy,
+					struct net_device *dev,
+					const u8 *peer, int link_id, u8 action_code,
+					u8 dialog_token, u16 status_code,
+					u32 peer_capability, bool initiator,
+					const u8 *buf, size_t len)
+#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 17, 0)) || defined(WITH_BACKPORTS)
 static int wlan_hdd_cfg80211_tdls_mgmt(struct wiphy *wiphy,
 					struct net_device *dev,
 					const u8 *peer, u8 action_code,
@@ -30516,7 +30523,10 @@ int wlan_hdd_cfg80211_send_tdls_discover_req(struct wiphy *wiphy,
     return wlan_hdd_cfg80211_tdls_mgmt(wiphy, dev, peer,
                             WLAN_TDLS_DISCOVERY_REQUEST, 1, 0, 0, NULL, 0);
 #else
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3,17,0)) || defined(WITH_BACKPORTS)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6,5,0)) || defined(WITH_BACKPORTS)
+    return wlan_hdd_cfg80211_tdls_mgmt(wiphy, dev, peer, 0,
+                            WLAN_TDLS_DISCOVERY_REQUEST, 1, 0, 0, 0, NULL, 0);
+#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(3,17,0)) || defined(WITH_BACKPORTS)
     return wlan_hdd_cfg80211_tdls_mgmt(wiphy, dev, peer,
                             WLAN_TDLS_DISCOVERY_REQUEST, 1, 0, 0, 0, NULL, 0);
 #elif (LINUX_VERSION_CODE >= KERNEL_VERSION(3,16,0))
diff --git a/CORE/MAC/src/include/parserApi.h b/CORE/MAC/src/include/parserApi.h
index c2fc46d..d2d3be1 100644
--- a/CORE/MAC/src/include/parserApi.h
+++ b/CORE/MAC/src/include/parserApi.h
@@ -38,7 +38,7 @@
 #ifndef __PARSE_H__
 #define __PARSE_H__
 
-#include <stdarg.h>
+#include <linux/stdarg.h>
 #include "sirMacPropExts.h"
 #include "dot11f.h"
 #ifdef WLAN_FEATURE_VOWIFI_11R
diff --git a/CORE/MAC/src/include/utilsApi.h b/CORE/MAC/src/include/utilsApi.h
index d81ee00..ae1d5d8 100644
--- a/CORE/MAC/src/include/utilsApi.h
+++ b/CORE/MAC/src/include/utilsApi.h
@@ -39,7 +39,7 @@
 #ifndef __UTILSAPI_H
 #define __UTILSAPI_H
 
-#include <stdarg.h>
+#include <linux/stdarg.h>
 #include <sirCommon.h>
 #include "aniGlobal.h"
 #include "utilsGlobal.h"
diff --git a/CORE/SME/src/csr/csrApiRoam.c b/CORE/SME/src/csr/csrApiRoam.c
index fd98f03..ef0daaf 100644
--- a/CORE/SME/src/csr/csrApiRoam.c
+++ b/CORE/SME/src/csr/csrApiRoam.c
@@ -9823,11 +9823,20 @@ void csrRoamJoinedStateMsgProcessor( tpAniSirGlobal pMac, void *pMsgBuf )
             tCsrRoamSession  *pSession;
             tSirSmeAssocIndToUpperLayerCnf *pUpperLayerAssocCnf;
             tCsrRoamInfo *roam_info;
-            tANI_U32 sessionId = CSR_ROAM_SESSION_MAX;
+            tANI_U32 sessionId;
             eHalStatus status;
             smsLog( pMac, LOG1, FL("ASSOCIATION confirmation can be given to upper layer "));
             pUpperLayerAssocCnf = (tSirSmeAssocIndToUpperLayerCnf *)pMsgBuf;
+
             status = csrRoamGetSessionIdFromBSSID( pMac, (tCsrBssid *)pUpperLayerAssocCnf->bssId, &sessionId );
+            if(!HAL_STATUS_SUCCESS(status))
+            {
+                smsLog(pMac, LOGE, FL(" Matching sessionId could not be found "));
+                if (pUpperLayerAssocCnf->ies)
+                    vos_mem_free(pUpperLayerAssocCnf->ies);
+                return;
+            }
+
             pSession = CSR_GET_SESSION(pMac, sessionId);
 
             if(!pSession)
@@ -17904,7 +17913,7 @@ eHalStatus csrGetSnr(tpAniSirGlobal pMac,
 {
    eHalStatus status = eHAL_STATUS_SUCCESS;
    vos_msg_t  msg;
-   tANI_U32 sessionId = CSR_ROAM_SESSION_MAX;
+   tANI_U32 sessionId;
 
    tAniGetSnrReq *pMsg;
 
diff --git a/CORE/SYS/legacy/src/system/inc/sysDebug.h b/CORE/SYS/legacy/src/system/inc/sysDebug.h
index 8cf24f9..0a023fa 100644
--- a/CORE/SYS/legacy/src/system/inc/sysDebug.h
+++ b/CORE/SYS/legacy/src/system/inc/sysDebug.h
@@ -37,7 +37,7 @@
 #ifndef __SYS_DEBUG_H__
 #define __SYS_DEBUG_H__
 
-#include <stdarg.h>
+#include <linux/stdarg.h>
 # include "utilsApi.h"
 # include "sirDebug.h"
 # include "sirParams.h"
diff --git a/CORE/SYS/legacy/src/utils/inc/utilsParser.h b/CORE/SYS/legacy/src/utils/inc/utilsParser.h
index aa2da85..0293aac 100644
--- a/CORE/SYS/legacy/src/utils/inc/utilsParser.h
+++ b/CORE/SYS/legacy/src/utils/inc/utilsParser.h
@@ -39,7 +39,7 @@
 #ifndef __UTILS_PARSE_H__
 #define __UTILS_PARSE_H__
 
-#include <stdarg.h>
+#include <linux/stdarg.h>
 #include "sirApi.h"
 #include "dot11f.h"
 #include "utilsApi.h"
diff --git a/CORE/SYS/legacy/src/utils/src/logApi.c b/CORE/SYS/legacy/src/utils/src/logApi.c
index 5763688..ba12329 100644
--- a/CORE/SYS/legacy/src/utils/src/logApi.c
+++ b/CORE/SYS/legacy/src/utils/src/logApi.c
@@ -41,7 +41,7 @@
 #include <utilsApi.h>
 #include <wlan_qct_wda.h>
 
-#include <stdarg.h>
+#include <linux/stdarg.h>
 #include "utilsGlobal.h"
 #include "macInitApi.h"
 #include "palApi.h"
diff --git a/CORE/UTILS/PKTLOG/linux_ac.c b/CORE/UTILS/PKTLOG/linux_ac.c
index 50a06b7..3c1c50c 100644
--- a/CORE/UTILS/PKTLOG/linux_ac.c
+++ b/CORE/UTILS/PKTLOG/linux_ac.c
@@ -339,6 +339,109 @@ ATH_SYSCTL_DECL(ath_sysctl_pktlog_size, ctl, write, filp, buffer, lenp,
 	return ret;
 }
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 5, 0))
+
+/* Register sysctl table */
+static int pktlog_sysctl_register(struct ol_softc *scn)
+{
+	struct ol_pktlog_dev_t *pl_dev = get_pl_handle(scn);
+	struct ath_pktlog_info_lnx *pl_info_lnx;
+	char *proc_name;
+
+	if (pl_dev) {
+		pl_info_lnx = PL_INFO_LNX(pl_dev->pl_info);
+		proc_name = pl_dev->name;
+	} else {
+		pl_info_lnx = PL_INFO_LNX(g_pktlog_info);
+		proc_name = PKTLOG_PROC_SYSTEM;
+	}
+
+#if (LINUX_VERSION_CODE > KERNEL_VERSION(2,6,31))
+#define set_ctl_name(a, b)	/* nothing */
+#else
+#define set_ctl_name(a, b)	pl_info_lnx->sysctls[a].ctl_name = b
+#endif
+
+	/*
+	 * Setup the sysctl table for creating the following sysctl entries:
+	 * /proc/sys/PKTLOG_PROC_DIR/<adapter>/enable for enabling/disabling
+	 * pktlog
+	 * /proc/sys/PKTLOG_PROC_DIR/<adapter>/size for changing the buffer size
+	 */
+	memset(pl_info_lnx->sysctls, 0, sizeof(pl_info_lnx->sysctls));
+
+
+	char tmp_dir_path[128];
+	snprintf(tmp_dir_path, sizeof(tmp_dir_path), "%s/%s", PKTLOG_PROC_DIR, proc_name);
+
+	pl_info_lnx->sysctls[0].procname = "enable";
+	pl_info_lnx->sysctls[0].mode = PKTLOG_PROCSYS_PERM;
+	pl_info_lnx->sysctls[0].proc_handler = ath_sysctl_pktlog_enable;
+	pl_info_lnx->sysctls[0].extra1 = scn;
+
+	pl_info_lnx->sysctls[1].procname = "size";
+	pl_info_lnx->sysctls[1].mode = PKTLOG_PROCSYS_PERM;
+	pl_info_lnx->sysctls[1].proc_handler = ath_sysctl_pktlog_size;
+	pl_info_lnx->sysctls[1].extra1 = scn;
+
+	pl_info_lnx->sysctls[2].procname = "options";
+	pl_info_lnx->sysctls[2].mode = PKTLOG_PROCSYS_PERM;
+	pl_info_lnx->sysctls[2].proc_handler = proc_dointvec;
+	pl_info_lnx->sysctls[2].data = &pl_info_lnx->info.options;
+	pl_info_lnx->sysctls[2].maxlen = sizeof(pl_info_lnx->info.options);
+
+	pl_info_lnx->sysctls[3].procname = "sack_thr";
+	pl_info_lnx->sysctls[3].mode = PKTLOG_PROCSYS_PERM;
+	pl_info_lnx->sysctls[3].proc_handler = proc_dointvec;
+	pl_info_lnx->sysctls[3].data = &pl_info_lnx->info.sack_thr;
+	pl_info_lnx->sysctls[3].maxlen = sizeof(pl_info_lnx->info.sack_thr);
+
+	pl_info_lnx->sysctls[4].procname = "tail_length";
+	pl_info_lnx->sysctls[4].mode = PKTLOG_PROCSYS_PERM;
+	pl_info_lnx->sysctls[4].proc_handler = proc_dointvec;
+	pl_info_lnx->sysctls[4].data = &pl_info_lnx->info.tail_length;
+	pl_info_lnx->sysctls[4].maxlen = sizeof(pl_info_lnx->info.tail_length);
+
+	pl_info_lnx->sysctls[5].procname = "thruput_thresh";
+	pl_info_lnx->sysctls[5].mode = PKTLOG_PROCSYS_PERM;
+	pl_info_lnx->sysctls[5].proc_handler = proc_dointvec;
+	pl_info_lnx->sysctls[5].data = &pl_info_lnx->info.thruput_thresh;
+	pl_info_lnx->sysctls[5].maxlen =
+		sizeof(pl_info_lnx->info.thruput_thresh);
+
+	pl_info_lnx->sysctls[6].procname = "phyerr_thresh";
+	pl_info_lnx->sysctls[6].mode = PKTLOG_PROCSYS_PERM;
+	pl_info_lnx->sysctls[6].proc_handler = proc_dointvec;
+	pl_info_lnx->sysctls[6].data = &pl_info_lnx->info.phyerr_thresh;
+	pl_info_lnx->sysctls[6].maxlen =
+		sizeof(pl_info_lnx->info.phyerr_thresh);
+
+	pl_info_lnx->sysctls[7].procname = "per_thresh";
+	pl_info_lnx->sysctls[7].mode = PKTLOG_PROCSYS_PERM;
+	pl_info_lnx->sysctls[7].proc_handler = proc_dointvec;
+	pl_info_lnx->sysctls[7].data = &pl_info_lnx->info.per_thresh;
+	pl_info_lnx->sysctls[7].maxlen = sizeof(pl_info_lnx->info.per_thresh);
+
+	pl_info_lnx->sysctls[8].procname = "trigger_interval";
+	pl_info_lnx->sysctls[8].mode = PKTLOG_PROCSYS_PERM;
+	pl_info_lnx->sysctls[8].proc_handler = proc_dointvec;
+	pl_info_lnx->sysctls[8].data = &pl_info_lnx->info.trigger_interval;
+	pl_info_lnx->sysctls[8].maxlen =
+		sizeof(pl_info_lnx->info.trigger_interval);
+	pl_info_lnx->sysctl_header =
+		register_sysctl(tmp_dir_path, pl_info_lnx->sysctls);
+
+
+	if (!pl_info_lnx->sysctl_header) {
+		printk("%s: failed to register sysctls!\n", proc_name);
+		return -1;
+	}
+
+	return 0;
+}
+
+#else
+
 /* Register sysctl table */
 static int pktlog_sysctl_register(struct ol_softc *scn)
 {
@@ -459,6 +562,8 @@ static int pktlog_sysctl_register(struct ol_softc *scn)
 	return 0;
 }
 
+#endif
+
 /*
  * Initialize logging for system or adapter
  * Parameter scn should be NULL for system wide logging
diff --git a/CORE/VOSS/inc/vos_trace.h b/CORE/VOSS/inc/vos_trace.h
index f5bc981..ad53386 100644
--- a/CORE/VOSS/inc/vos_trace.h
+++ b/CORE/VOSS/inc/vos_trace.h
@@ -44,7 +44,7 @@
   Include Files
   ------------------------------------------------------------------------*/
 #include  <vos_types.h>  // For VOS_MODULE_ID...
-#include  <stdarg.h>       // For va_list...
+#include  <linux/stdarg.h>       // For va_list...
 #include  <vos_status.h>
 #include  <i_vos_types.h>
 
diff --git a/CORE/VOSS/src/vos_utils.c b/CORE/VOSS/src/vos_utils.c
index 43a5e69..fb94f01 100644
--- a/CORE/VOSS/src/vos_utils.c
+++ b/CORE/VOSS/src/vos_utils.c
@@ -696,7 +696,7 @@ int hmac_sha1(v_U8_t *key, v_U8_t ksize, char *plaintext, v_U8_t psize,
     }
 
     ahash_request_set_callback(req, CRYPTO_TFM_REQ_MAY_BACKLOG,
-                               hmac_sha1_complete, &tresult);
+                               (void *)hmac_sha1_complete, &tresult);
 
     hash_buff = kzalloc(psize, GFP_KERNEL);
     if (!hash_buff) {
@@ -869,7 +869,7 @@ int hmac_md5(v_U8_t *key, v_U8_t ksize, char *plaintext, v_U8_t psize,
     }
 
     ahash_request_set_callback(req, CRYPTO_TFM_REQ_MAY_BACKLOG,
-                                        hmac_md5_complete, &tresult);
+                                        (void *)hmac_md5_complete, &tresult);
 
     hash_buff = kzalloc(psize, GFP_KERNEL);
     if (!hash_buff) {
diff --git a/CORE/WDA/src/wlan_qct_wda_debug.h b/CORE/WDA/src/wlan_qct_wda_debug.h
index d741d7b..d42b361 100644
--- a/CORE/WDA/src/wlan_qct_wda_debug.h
+++ b/CORE/WDA/src/wlan_qct_wda_debug.h
@@ -30,7 +30,7 @@
 #if  !defined (ANI_OS_TYPE_ANDROID)
 #include <stdio.h>
 #endif
-#include <stdarg.h>
+#include <linux/stdarg.h>
 
 #include "utilsApi.h"
 #include "sirDebug.h"
-- 
2.14.5

